File: ios/Stash/Model/Protocol.swift
import Foundation

/// Messages received from the server. Mirrors Rust `protocol::ServerMsg`
/// (serde tagged enum, `type` discriminator, snake_case).
enum ServerMessage {
    case challenge(nonce: String)
    case authOk
    case authFail
    case syncMeta(items: [SyncItem])
    case itemAdded(item: SyncItem)
    case itemDeleted(id: String)
    case fileChunk(id: String, chunk: String, index: Int, total: Int)
    case fileTransferComplete(id: String)
    case fileVerifyFailed(id: String)
    case uploadState(id: String, received: Int)
    case fileUnavailable(id: String)
    case error(message: String)
    case unknown

    static func parse(_ text: String) -> ServerMessage {
        guard let data = text.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = obj["type"] as? String
        else { return .unknown }

        func str(_ k: String) -> String { obj[k] as? String ?? "" }
        func int(_ k: String) -> Int { (obj[k] as? NSNumber)?.intValue ?? 0 }

        switch type {
        case "challenge": return .challenge(nonce: str("nonce"))
        case "auth_ok": return .authOk
        case "auth_fail": return .authFail
        case "sync_meta":
            let items = decodeItems(obj["items"])
            return .syncMeta(items: items)
        case "item_added":
            guard let item = decodeItem(obj["item"]) else { return .unknown }
            return .itemAdded(item: item)
        case "item_deleted": return .itemDeleted(id: str("id"))
        case "file_chunk":
            return .fileChunk(id: str("id"), chunk: str("chunk"), index: int("index"), total: int("total"))
        case "file_transfer_complete": return .fileTransferComplete(id: str("id"))
        case "file_verify_failed": return .fileVerifyFailed(id: str("id"))
        case "upload_state": return .uploadState(id: str("id"), received: int("received"))
        case "file_unavailable": return .fileUnavailable(id: str("id"))
        case "error": return .error(message: str("message"))
        default: return .unknown
        }
    }

    private static func decodeItem(_ any: Any?) -> SyncItem? {
        guard let dict = any,
              let data = try? JSONSerialization.data(withJSONObject: dict) else { return nil }
        return try? JSONDecoder().decode(SyncItem.self, from: data)
    }

    private static func decodeItems(_ any: Any?) -> [SyncItem] {
        guard let arr = any,
              let data = try? JSONSerialization.data(withJSONObject: arr) else { return [] }
        return (try? JSONDecoder().decode([SyncItem].self, from: data)) ?? []
    }
}

/// Messages sent to the server. Mirrors Rust `protocol::ClientMsg`.
enum ClientMessage {
    case auth(hash: String, knownIds: [String])
    case syncMeta(items: [SyncItem])
    case pushItem(item: SyncItem)
    case deleteItem(id: String)
    case requestFile(id: String, fromChunk: Int)
    case pushFileChunk(id: String, chunk: String, index: Int, total: Int)
    case queryUploadState(id: String, total: Int)

    /// Serialize to the JSON text frame the server expects.
    func json() -> String {
        var obj: [String: Any]
        switch self {
        case let .auth(hash, knownIds):
            obj = ["type": "auth", "hash": hash]
            if !knownIds.isEmpty { obj["known_ids"] = knownIds }
        case let .syncMeta(items):
            obj = ["type": "sync_meta", "items": items.map { $0.asDictionary() }]
        case let .pushItem(item):
            obj = ["type": "push_item", "item": item.asDictionary()]
        case let .deleteItem(id):
            obj = ["type": "delete_item", "id": id]
        case let .requestFile(id, fromChunk):
            obj = ["type": "request_file", "id": id]
            if fromChunk > 0 { obj["from_chunk"] = fromChunk }
        case let .pushFileChunk(id, chunk, index, total):
            obj = ["type": "push_file_chunk", "id": id, "chunk": chunk, "index": index, "total": total]
        case let .queryUploadState(id, total):
            obj = ["type": "query_upload_state", "id": id, "total": total]
        }
        guard let data = try? JSONSerialization.data(withJSONObject: obj),
              let s = String(data: data, encoding: .utf8) else { return "{}" }
        return s
    }
}

private extension SyncItem {
    /// Encode to a `[String: Any]` with the exact snake_case keys the server reads.
    func asDictionary() -> [String: Any] {
        var d: [String: Any] = [
            "id": id,
            "title": title,
            "content": content,
            "type": type,
            "mime_type": mimeType,
            "is_pinned": isPinned,
            "created_at": createdAt,
            "updated_at": updatedAt,
            "has_file": hasFile,
            "file_hash": fileHash,
        ]
        if !thumbnailB64.isEmpty { d["thumbnail_b64"] = thumbnailB64 }
        return d
    }
}

File: server/src/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMsg {
    Challenge {
        nonce: String,
    },
    AuthOk,
    AuthFail,
    SyncMeta {
        items: Vec<Item>,
    },
    ItemAdded {
        item: Item,
    },
    ItemDeleted {
        id: String,
    },
    FileChunk {
        id: String,
        chunk: String,
        index: u32,
        total: u32,
    },
    FileTransferComplete {
        id: String,
    },
    /// Sent when a fully-received file's SHA-256 does not match the item's recorded
    /// file_hash. The server discards the file; the client should re-upload.
    FileVerifyFailed {
        id: String,
    },
    /// How many consecutive chunks (from index 0) the server already has for an upload.
    UploadState {
        id: String,
        received: u32,
    },
    /// The item exists but its file isn't on this server yet (e.g. a mirror hasn't pulled it).
    /// Transient + keyed to the id so the client can retry just that download.
    FileUnavailable {
        id: String,
    },
    Error {
        message: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMsg {
    /// hash: SHA-256 hex of the password. known_ids: item syncIds the client already has
    /// (server omits thumbnail_b64 for these in the SyncMeta response).
    Auth {
        hash: String,
        #[serde(default)]
        known_ids: Vec<String>,
    },
    SyncMeta {
        items: Vec<Item>,
    },
    PushItem {
        item: Item,
    },
    DeleteItem {
        id: String,
    },
    /// from_chunk: resume download from this chunk index (0 = full download).
    RequestFile {
        id: String,
        #[serde(default)]
        from_chunk: u32,
    },
    PushFileChunk {
        id: String,
        chunk: String,
        index: u32,
        total: u32,
    },
    /// Ask how many consecutive chunks the server already has for a pending upload.
    /// `total` is the uploader's chunk count; if it doesn't match the server's saved
    /// partial (e.g. the chunk size changed between versions), the server restarts from 0.
    QueryUploadState {
        id: String,
        #[serde(default)]
        total: u32,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Item {
    pub id: String,
    pub title: String,
    pub content: String,
    #[serde(rename = "type")]
    pub item_type: String,
    pub mime_type: String,
    pub is_pinned: bool,
    pub created_at: i64,
    /// Last-modification timestamp (epoch millis). Drives last-writer-wins on edits.
    /// Defaults to 0 for older clients; treated as created_at in that case.
    #[serde(default)]
    pub updated_at: i64,
    pub has_file: bool,
    #[serde(default)]
    pub file_hash: String,
    #[serde(default)]
    pub thumbnail_b64: String,
}

impl Item {
    /// Effective modification time — falls back to created_at for clients that
    /// don't send updated_at yet.
    pub fn effective_updated_at(&self) -> i64 {
        self.updated_at.max(self.created_at)
    }
}
